README.md
# adsorptionview

[![pub package](https://img.shields.io/pub/v/adsorptionview_flutter.svg)](https://pub.dartlang.org/packages/adsorptionview_flutter)

ListView adsorption ceiling layout, only support Equally high items.

ListView的吸顶布局，现在仅支持等高的Item。

HomePage：[https://github.com/baoolong/Adsorptionview](https://github.com/baoolong/Adsorptionview)

MoreWidght：[https://github.com/OpenFlutter/PullToRefresh](https://github.com/OpenFlutter/PullToRefresh)

<img width="38%" height="38%" src="https://raw.githubusercontent.com/OpenFlutter/PullToRefresh/master/demonstrationgif/20180912_100745.gif"/>

## Usage

Add this to your package's pubspec.yaml file:

	dependencies:
	  adsorptionview_flutter: ^0.1.3
	  
Add it to your dart file:

    import 'package:adsorptionview_flutter/adsorptionview_flutter.dart';
    
And ListView dataBin must extends AdsorptionData ,Add it to your dataBin file
    
    import 'package:adsorptionview_flutter/adsorptiondatabin.dart';

## Example

#### 	Bin

    import 'package:adsorptionview_flutter/adsorptiondatabin.dart';
   
    class AdsorptionListBin extends AdsorptionData{
      AdsorptionListBin( this.headerName);
   
      String headerName;
   
      @override
      String toString() {
        return 'AdsorptionListBin{headerName: $headerName}';
      }
   
    }
   
   
#### 	Demo
    import 'package:flutter/material.dart';
    import 'package:flutter/rendering.dart';
    import 'package:flutterapp/bin/adsorptionlistbin.dart';
    import 'package:adsorptionview_flutter/adsorptionview_flutter.dart';
    
    class AdsorptionViewDemo extends StatefulWidget{
      @override
      State<StatefulWidget> createState() {
        List<AdsorptionListBin> adsorptionDatas=new List();
        AdsorptionListBin adsorptionData;
    
        adsorptionData=new AdsorptionListBin("A");
        adsorptionData.isHeader=true;
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("阿杜");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("阿宝");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("艾夫杰尼");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("阿牛");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("安苏羽");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("阿勒长青");
        adsorptionDatas.add(adsorptionData);
    
    
        adsorptionData=new AdsorptionListBin("B");
        adsorptionData.isHeader=true;
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("白小白");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("白羽毛");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("Bridge");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("斑马");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("白一阳");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("白举纲");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("暴林");
        adsorptionDatas.add(adsorptionData);
    
    
        adsorptionData=new AdsorptionListBin("C");
        adsorptionData.isHeader=true;
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("陈奕迅");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("陈小春");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("成龙");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("陈百强");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("迟志强");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("崔健");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("陈晓东");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("陈学冬");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("蔡国庆");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("陈冠希");
        adsorptionDatas.add(adsorptionData);
    
        adsorptionData=new AdsorptionListBin("陈琳");
        adsorptionDatas.add(adsorptionData);
        return new AdsorptionViewState(adsorptionDatas);
      }
    }
    
    ///此控件适用于固定高度的ListView
    class AdsorptionViewState extends State<AdsorptionViewDemo>{
    
      AdsorptionViewState(this.adsorptionDatas);
    
      List<AdsorptionListBin> adsorptionDatas;
      double itemHeight=50.0;
    
    
      @override
      Widget build(BuildContext context) {
        return new Scaffold(
          appBar: new AppBar(
            title: new Text("吸附布局"),
          ),
          body:new AdsorptionView(
            adsorptionDatas: adsorptionDatas,
            generalItemChild: (AdsorptionListBin bin) {
              return new Container(
                alignment: Alignment.centerLeft,
                margin: EdgeInsets.fromLTRB(15.0, 0.0, 0.0, 0.0),
                child: new Text(
                  bin.headerName,
                  style: new TextStyle(fontSize: 18.0, color: Colors.grey),
                ),
              );
            },
            headChild: (AdsorptionListBin bin) {
              return new Container(
                color: Colors.grey,
                alignment: Alignment.centerLeft,
                padding: EdgeInsets.fromLTRB(15.0, 0.0, 0.0, 0.0),
                child: new Text(
                  bin.headerName,
                  style: new TextStyle( fontSize: 20.0,color: Colors.black),
                ),
              );
            },
          ),
        );
      }
    }

## Notice

吸顶布局不能使用类似苹果的弹簧布局，以后可能会支持；目前等高的支持比较好，非等高的有点问题，后续会改进

Adsorptionview ceiling layout can not use the spring layout of Apple, may be supported in the future; the current support of the equally hight items better, the  unequally hight items  has a problem, the follow-up will improve

## LICENSE
    MIT License

	Copyright (c) 2018 baoolong
	
	Permission is hereby granted, free of charge, to any person obtaining a copy
	of this software and associated documentation files (the "Software"), to deal
	in the Software without restriction, including without limitation the rights
	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
	copies of the Software, and to permit persons to whom the Software is
	furnished to do so, subject to the following conditions:
	
	The above copyright notice and this permission notice shall be included in all
	copies or substantial portions of the Software.
	
	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
	SOFTWARE.


CHANGELOG.md
## [0.1.3] - 2018.10.18

*   optimization and Non-equal height items are no longer supported

## [0.1.2] - 2018.10.18

*   Prevent IOS devices from springing

## [0.1.1] - 2018.10.15.

*   Describe initial release.


test/adsorptionlistbin.dart
import 'package:adsorptionview_flutter/adsorptiondatabin.dart';

class AdsorptionListBin extends AdsorptionData{
  AdsorptionListBin( this.headerName);

  String headerName;

  @override
  String toString() {
    return 'AdsorptionListBin{headerName: $headerName}';
  }

}

test/adsorptionview_test.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:adsorptionview_flutter/adsorptionview_flutter.dart';
import 'adsorptionlistbin.dart';

void main() => runApp(
  new MaterialApp(
    title: 'AdsorptionViewDemo',
    theme: new ThemeData(
      primaryColor: Colors.white,
    ),
    home: new AdsorptionViewDemo(),
  )
);



class AdsorptionViewDemo extends StatefulWidget{
  @override
  State<StatefulWidget> createState() {
    List<AdsorptionListBin> adsorptionDatas=new List();
    AdsorptionListBin adsorptionData;

    adsorptionData=new AdsorptionListBin("A");
    adsorptionData.isHeader=true;
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("阿杜");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("阿宝");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("艾夫杰尼");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("阿牛");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("安苏羽");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("阿勒长青");
    adsorptionDatas.add(adsorptionData);


    adsorptionData=new AdsorptionListBin("B");
    adsorptionData.isHeader=true;
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("白小白");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("白羽毛");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("Bridge");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("斑马");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("白一阳");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("白举纲");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("暴林");
    adsorptionDatas.add(adsorptionData);


    adsorptionData=new AdsorptionListBin("C");
    adsorptionData.isHeader=true;
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈奕迅");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈小春");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("成龙");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈百强");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("迟志强");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("崔健");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈晓东");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈学冬");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("蔡国庆");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈冠希");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈琳");
    adsorptionDatas.add(adsorptionData);
    return new AdsorptionViewState(adsorptionDatas);
  }
}

///此控件适用于固定高度的ListView
class AdsorptionViewState extends State<AdsorptionViewDemo>{

  AdsorptionViewState(this.adsorptionDatas);

  List<AdsorptionListBin> adsorptionDatas;
  double itemHeight=50.0;


  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text("吸附布局"),
      ),
      body:new AdsorptionView(
        isEqualHeightItem: true,
        adsorptionDatas: adsorptionDatas,
        generalItemChild: (AdsorptionListBin bin) {
          return new Container(
            alignment: Alignment.centerLeft,
            margin: EdgeInsets.fromLTRB(15.0, 0.0, 0.0, 0.0),
            child: new Text(
              bin.headerName,
              style: new TextStyle(fontSize: 18.0, color: Colors.grey),
            ),
          );
        },
        headChild: (AdsorptionListBin bin) {
          return new Container(
            color: Colors.grey,
            alignment: Alignment.centerLeft,
            padding: EdgeInsets.fromLTRB(15.0, 0.0, 0.0, 0.0),
            child: new Text(
              bin.headerName,
              style: new TextStyle( fontSize: 20.0,color: Colors.black),
            ),
          );
        },
      ),
    );
  }

}

lib/adsorptionview_flutter.dart
import 'package:adsorptionview_flutter/adsorptiondatabin.dart';
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';

typedef Widget GetHearWidget <M extends AdsorptionData> (M bin);
typedef Widget GetGeneralItem <M extends AdsorptionData> (M bin);

class AdsorptionView<T extends AdsorptionData> extends StatefulWidget{

  final List<T> adsorptionDatas;
  final GetHearWidget<T> headChild;
  final GetGeneralItem<T> generalItemChild;
  final double itemHeight;
  final double itemWidth;
  final double cacheExtent;

  AdsorptionView({
    @required this.adsorptionDatas,
    @required this.headChild,
    @required this.generalItemChild,
    this.itemHeight:50.0,
    this.itemWidth:double.infinity,
    this.cacheExtent:30.0,
  }): assert(
  adsorptionDatas!=null,
  generalItemChild!=null&&
      headChild!=null,
  );

  @override
  State<StatefulWidget> createState() {
    return new AdsorptionViewState<T>();
  }
}

///此控件适用于固定高度的ListView
class AdsorptionViewState<T extends AdsorptionData> extends State<AdsorptionView<T>>{

  ScrollController scrollController=new ScrollController();

  @override
  Widget build(BuildContext context) {
    return new Stack(
      children: <Widget>[
        new AdsorptionListView(
          scrollController: scrollController,
          adsorptionDatas: widget.adsorptionDatas,
          generalItemChild: widget.generalItemChild,
          headChild: widget.headChild,
          itemHeight: widget.itemHeight,
          itemWidth: widget.itemWidth,
          cacheExtent: widget.cacheExtent,
        ),
        new GestureDetector(
          onTap: (){
            double pixels= scrollController.position.pixels;
            int a=pixels~/widget.itemHeight;
            for(int i=a;i>=0;i--){
              if(widget.adsorptionDatas[i].isHeader) {
                scrollController.animateTo(i*widget.itemHeight, duration: new Duration(milliseconds: 200), curve: Curves.linear);
                break;
              }
            }
          },
          child: new HeaderView(
            scrollController: scrollController,
            headChild: widget.headChild,
            adsorptionDatas: widget.adsorptionDatas,
            itemWidth: widget.itemWidth,
            itemHeight: widget.itemHeight,
          ),
        ),
      ],
    );
  }
}


class AdsorptionListView<T extends AdsorptionData> extends StatefulWidget{
  final ScrollController scrollController;
  final double itemHeight;
  final double itemWidth;
  final double cacheExtent;
  final List<T> adsorptionDatas;
  final GetHearWidget<T> headChild;
  final GetGeneralItem<T> generalItemChild;

  AdsorptionListView({
    @required this.adsorptionDatas,
    @required this.headChild,
    @required this.generalItemChild,
    this.itemHeight:50.0,
    this.itemWidth:double.infinity,
    this.cacheExtent:30.0,
    @required this.scrollController,
  }): assert(
  adsorptionDatas!=null,
  generalItemChild!=null&&
      headChild!=null,
  );

  @override
  State<StatefulWidget> createState() {
    return AdsorptionListViewState<T>();
  }
}

class AdsorptionListViewState<T extends AdsorptionData> extends State<AdsorptionListView<T>>{

  ScrollPhysics scrollPhysics=new ClampingScrollPhysics();

  @override
  Widget build(BuildContext context) {
    return new ListView.builder(
      physics: scrollPhysics,
      cacheExtent: widget.cacheExtent,
      controller: widget.scrollController,
      itemCount: widget.adsorptionDatas.length,
      itemBuilder: (context, index) {
        if(widget.adsorptionDatas[index].isHeader){
          return new Container(
            width: widget.itemWidth,
            height: widget.itemHeight,
            child: widget.headChild(widget.adsorptionDatas[index]),
          );
        }else{
          return new Container(
            width: widget.itemWidth,
            height: widget.itemHeight,
            child: widget.generalItemChild(widget.adsorptionDatas[index]),
          );
        }
      },
    );
  }
}


class HeaderView<T extends AdsorptionData> extends StatefulWidget{

  final ScrollController scrollController;
  final double itemHeight;
  final double itemWidth;
  final GetHearWidget<T> headChild;
  final List<T> adsorptionDatas;

  HeaderView({
    @required this.scrollController,
    this.itemHeight:50.0,
    this.itemWidth:double.infinity,
    @required this.headChild,
    @required this.adsorptionDatas,
  });

  @override
  State<StatefulWidget> createState() {
    return new HeaderViewState<T>();
  }
}

class HeaderViewState<T extends AdsorptionData> extends State<HeaderView<T>>{
  double headerOffset=0.0;
  T headerStr;
  double beforeScroll=0.0;

  @override
  void initState() {
    headerStr=widget.adsorptionDatas.first;

    widget.scrollController.addListener((){
      //计算滑动了多少距离了
      double pixels=widget.scrollController.position.pixels;

      //根据滑动的距离 计算当前可见的第一个Item的Position
      int a=pixels~/widget.itemHeight;
      //计算滑动出屏幕多少距离
      double b=pixels%widget.itemHeight;
      double currentScrollPosition=widget.scrollController.position.extentBefore;
      //如果下一个item是Header 则偏移 如果不是 则偏移量=0
      if(widget.adsorptionDatas[a+1].isHeader){
        setState(() {
          // 改变布局
          if(currentScrollPosition-beforeScroll<0){
            //检测到再向上划就越出当前组 提前改变header的内容并偏移
            for(int i=a;i>=0;i--){
              if(widget.adsorptionDatas[i].isHeader){
                headerStr=widget.adsorptionDatas[i];
                break;
              }
            }
          }
          beforeScroll=currentScrollPosition;
          headerOffset=-b;
        });
      }else{
        //始终使header处于完整显示状态
        for(int i=a;i>=0;i--){
          if(widget.adsorptionDatas[i].isHeader) {
            if(headerStr != widget.adsorptionDatas[i]) {
              setState(() {
                headerStr = widget.adsorptionDatas[i];
              });
            }
            break;
          }
        }
        if(headerOffset!=0){
          setState(() {
            headerOffset=0.0;
          });
        }
      }
    });
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return new Container(
      transform: Matrix4.translationValues(0.0, headerOffset, 0.0),
      width: widget.itemWidth,
      height: widget.itemHeight,
      child: widget.headChild(headerStr),
    );
  }
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'package:adsorptionview_flutter/adsorptionview_flutter.dart';
import 'adsorptionlistbin.dart';

void main() => runApp(
    new MaterialApp(
      title: 'AdsorptionViewDemo',
      theme: new ThemeData(
        primaryColor: Colors.white,
      ),
      home: new AdsorptionViewDemo(),
    )
);



class AdsorptionViewDemo extends StatefulWidget{
  @override
  State<StatefulWidget> createState() {
    List<AdsorptionListBin> adsorptionDatas=new List();
    AdsorptionListBin adsorptionData;

    adsorptionData=new AdsorptionListBin("A");
    adsorptionData.isHeader=true;
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("阿杜");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("阿宝");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("艾夫杰尼");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("阿牛");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("安苏羽");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("阿勒长青");
    adsorptionDatas.add(adsorptionData);


    adsorptionData=new AdsorptionListBin("B");
    adsorptionData.isHeader=true;
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("白小白");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("白羽毛");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("Bridge");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("斑马");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("白一阳");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("白举纲");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("暴林");
    adsorptionDatas.add(adsorptionData);


    adsorptionData=new AdsorptionListBin("C");
    adsorptionData.isHeader=true;
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈奕迅");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈小春");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("成龙");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈百强");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("迟志强");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("崔健");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈晓东");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈学冬");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("蔡国庆");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈冠希");
    adsorptionDatas.add(adsorptionData);

    adsorptionData=new AdsorptionListBin("陈琳");
    adsorptionDatas.add(adsorptionData);
    return new AdsorptionViewState(adsorptionDatas);
  }
}

///此控件适用于固定高度的ListView
class AdsorptionViewState extends State<AdsorptionViewDemo>{

  AdsorptionViewState(this.adsorptionDatas);

  List<AdsorptionListBin> adsorptionDatas;
  double itemHeight=50.0;


  @override
  Widget build(BuildContext context) {
    return new Scaffold(
      appBar: new AppBar(
        title: new Text("吸附布局"),
      ),
      body:new AdsorptionView(
        adsorptionDatas: adsorptionDatas,
        generalItemChild: (AdsorptionListBin bin) {
          return new Container(
            alignment: Alignment.centerLeft,
            margin: EdgeInsets.fromLTRB(15.0, 0.0, 0.0, 0.0),
            child: new Text(
              bin.headerName,
              style: new TextStyle(fontSize: 18.0, color: Colors.grey),
            ),
          );
        },
        headChild: (AdsorptionListBin bin) {
          return new Container(
            color: Colors.grey,
            alignment: Alignment.centerLeft,
            padding: EdgeInsets.fromLTRB(15.0, 0.0, 0.0, 0.0),
            child: new Text(
              bin.headerName,
              style: new TextStyle( fontSize: 20.0,color: Colors.black),
            ),
          );
        },
      ),
    );
  }

}